### PR TITLE
Profile: On dataChanged() only update pictures that actually changed

### PR DESCRIPTION
**Pull request by bstoeger**
Only update those pictures of the DivePictureModel that actually changed.
This will be useful once pictures are loaded incrementally.

To do so, replace the pictures array by an array with stable ids. Before
this commit, not-shown pictures are left out of the pictures array, which
makes the mapping from DivePictureModel-ids to the picture array index
non-trivial.

Replace the QList<DivePictureItem *> by a std::vector<std::unique_ptr<DivePictureItem>>
to easy memory management. Sadly, owing to COW semantics, QVector is incompatible
with QScopedPointer.

Signed-off-by: Berthold Stoeger <bstoeger@mail.tuwien.ac.at>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

Only update changed pictures in the profile plot. For discussion, see #1179.

### CODE REVIEW

**Review by atdotde — COMMENTED** [review]
I admit I am uneasy with adding another source of complex data types (std::xxx in addition to Qt types). Obviously, it's my lacking of understanding of C++, but I guess (if I understand correctly what is going on) I would indirect rather via a QHash. In any case, I would like to hear @tcanabrava or @thiagomacieira  on this.

File: profile-widget/profilewidget2.cpp [hunk 1]
```diff
@@ -1103,7 +1106,7 @@ void ProfileWidget2::setProfileState()
 
 	disconnectTemporaryConnections();
 #ifndef SUBSURFACE_MOBILE
-	connect(DivePictureModel::instance(), SIGNAL(dataChanged(QModelIndex, QModelIndex)), this, SLOT(plotPictures()));
+	connect(DivePictureModel::instance(), &DivePictureModel::dataChanged, this, &ProfileWidget2::updatePictures);
```
Comment (thiagomacieira) [inline profile-widget/profilewidget2.cpp]: Unrelated code clean up?

File: profile-widget/profilewidget2.cpp [hunk 2]
```diff
@@ -1974,26 +1977,46 @@ void ProfileWidget2::keyEscAction()
...
+		if (!offsetSeconds) {
+			pictures[picItemNr].reset();
 			continue;
-		DivePictureItem *item = new DivePictureItem();
+		}
+		if (!pictures[picItemNr]) {
+			pictures[picItemNr].reset(new DivePictureItem);
+			scene()->addItem(pictures[picItemNr].get());
+		}
+		DivePictureItem *item = pictures[picItemNr].get();
```
Comment (thiagomacieira) [inline profile-widget/profilewidget2.cpp]: Suggestion: if you move this line before the `if`, you can save yourself from repeating `pictures[picItemNr]` three times. It's ok to `.get()` a null unique pointer -- you get nullptr.

**Comment by thiagomacieira** [comment]
Code looks good, I don't see any problems in the change. I don't know the rest of the code to say whether there would be problems with the parts that aren't changed, but it's unlikely.

COW semantics made sense in the 90s when Qt started, before move semantics existed. There's also the fact that most developers aren't as good as you are, so providing it results in better code for an average developer. Those who know how to use `std::vector` properly like you can use it.

I'll also defend UTF-16. It's a lot easier to manipulate than UTF-8, but doesn't require as much space as UCS-4. Microsoft still recommends it and all the major APIs use it: ICU, Java, Cocoa and Win32.

**Comment by dirkhh** [comment]
OK, this appears to have gotten the seal of approval, merging it now
